From 99a644a82eea67f48bf931025c431b77cac58e9b Mon Sep 17 00:00:00 2001
From: Alberto Milone <alberto.milone@canonical.com>
Date: Tue, 7 Jan 2020 14:23:13 +0100
Subject: [PATCH 1/3] Add support for Linux 5.2

V3: drop the DRIVER_LEGACY part which is now upstream
    keep the _fault detection code.
---
 uvm/nvidia_uvm_lite.c | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/uvm/nvidia_uvm_lite.c b/uvm/nvidia_uvm_lite.c
index 87eefac..fd7a4ce 100644
--- a/uvm/nvidia_uvm_lite.c
+++ b/uvm/nvidia_uvm_lite.c
@@ -691,7 +691,11 @@ static UvmGpuMigrationTracking * _get_mig_tracker(UvmCommitRecord * pRecord)
 //         return SIGBUS.
 //      3. Otherwise, map in a page from the cache, and allow access.
 //
+#ifdef VM_FAULT_OOM
 int _fault_common(struct vm_area_struct *vma, unsigned long vaddr,
+#else
+unsigned int _fault_common(struct vm_area_struct *vma, unsigned long vaddr,
+#endif
                   struct page **ppage, unsigned vmfFlags)
 {
     int retValue = VM_FAULT_SIGBUS;
@@ -821,9 +825,17 @@ done:
 
 #if defined(NV_VM_OPERATIONS_STRUCT_HAS_FAULT)
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4, 11, 0)
+#ifdef VM_FAULT_OOM
 int _fault(struct vm_area_struct *vma, struct vm_fault *vmf)
 #else
+unsigned int _fault(struct vm_area_struct *vma, struct vm_fault *vmf)
+#endif
+#else
+#ifdef VM_FAULT_OOM
 int _fault(struct vm_fault *vmf)
+#else
+unsigned int _fault(struct vm_fault *vmf)
+#endif
 #endif
 {
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 11, 0)
@@ -876,9 +888,17 @@ static struct vm_operations_struct uvmlite_vma_ops =
 //
 #if defined(NV_VM_OPERATIONS_STRUCT_HAS_FAULT)
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4, 11, 0)
+#ifdef VM_FAULT_OOM
 int _sigbus_fault(struct vm_area_struct *vma, struct vm_fault *vmf)
 #else
+unsigned int _sigbus_fault(struct vm_area_struct *vma, struct vm_fault *vmf)
+#endif
+#else
+#ifdef VM_FAULT_OOM
 int _sigbus_fault(struct vm_fault *vmf)
+#else
+unsigned int _sigbus_fault(struct vm_fault *vmf)
+#endif
 #endif
 {
     vmf->page = NULL;
-- 
2.20.1

